File: src/main/scala/resonant/lib/content/module/BlockDummy.java
package resonant.lib.content.module;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import net.minecraft.block.Block;
import net.minecraft.block.ITileEntityProvider;
import net.minecraft.client.renderer.texture.IconRegister;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.Icon;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import resonant.lib.content.module.TileBlock.IComparatorInputOverride;
import resonant.lib.prefab.vector.Cuboid;
import resonant.lib.render.block.BlockRenderingHandler;
import resonant.lib.utility.inventory.InventoryUtility;
import universalelectricity.api.vector.Vector3;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

public class BlockDummy extends Block implements ITileEntityProvider
{
    /** A dummy instance of the block used to forward methods to. */
    public final TileBlock dummyTile;

    public BlockDummy(int id, String modPrefix, CreativeTabs defaultTab, TileBlock dummyTile)
    {
        super(id, dummyTile.material);
        this.dummyTile = dummyTile;
        dummyTile.domain = modPrefix;
        setUnlocalizedName(modPrefix + dummyTile.name);
        setTextureName(modPrefix + dummyTile.textureName);

        if (dummyTile.creativeTab != null)
        {
            setCreativeTab(dummyTile.creativeTab);
        }
        else
        {
            setCreativeTab(defaultTab);
        }

        dummyTile.bounds.setBounds(this);

        /** Reinject opaqueCube data */
        opaqueCubeLookup[id] = isOpaqueCube();
        lightOpacity[id] = isOpaqueCube() ? 255 : 0;

        setHardness(dummyTile.blockHardness);
        setResistance(dummyTile.blockResistance);
        setTickRandomly(dummyTile.tickRandomly);
    }

    /** Injects and ejects data from the TileEntity. */
    public void inject(IBlockAccess access, int x, int y, int z)
    {
        if (access instanceof World)
        {
            dummyTile.worldObj = (World) access;
        }

        dummyTile.access = access;
        dummyTile.xCoord = x;
        dummyTile.yCoord = y;
        dummyTile.zCoord = z;

        TileEntity tile = access.getBlockTileEntity(x, y, z);

        if (tile instanceof TileBlock)
        {
            ((TileBlock) tile).block = this;
        }
    }

    public void eject()
    {
        dummyTile.worldObj = null;
        dummyTile.xCoord = 0;
        dummyTile.yCoord = 0;
        dummyTile.zCoord = 0;
    }

    public TileBlock getTile(IBlockAccess world, int x, int y, int z)
    {
        TileEntity tile = world.getBlockTileEntity(x, y, z);

        if (tile instanceof TileBlock)
        {
            return (TileBlock) tile;
        }

        return dummyTile;
    }

    @Override
    public boolean hasTileEntity(int metadata)
    {
        return dummyTile.tile() != null;
    }

    @Override
    public TileEntity createNewTileEntity(World world)
    {
        try
        {
            return dummyTile.tile().getClass().newInstance();
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public void fillWithRain(World world, int x, int y, int z)
    {
        inject(world, x, y, z);
        getTile(world, x, y, z).onFillRain();
        eject();
    }

    @Override
    public void onBlockClicked(World world, int x, int y, int z, EntityPlayer player)
    {
        inject(world, x, y, z);
        // TODO: Raytrace player's look position to determine the hit.
        getTile(world, x, y, z).click(player);
        eject();
    }

    @Override
    public void onBlockAdded(World world, int x, int y, int z)
    {
        inject(world, x, y, z);
        getTile(world, x, y, z).onAdded();
        eject();
    }

    @Override
    public void onBlockPlacedBy(World world, int x, int y, int z, EntityLivingBase entityLiving, ItemStack itemStack)
    {
        inject(world, x, y, z);
        getTile(world, x, y, z).onPlaced(entityLiving, itemStack);
        eject();
    }

    @Override
    public void breakBlock(World world, int x, int y, int z, int par5, int par6)
    {
        inject(world, x, y, z);
        getTile(world, x, y, z).onRemove(par5, par6);
        eject();
        super.breakBlock(world, x, y, z, par5, par6);
    }

    @Override
    public int quantityDropped(int meta, int fortune, Random random)
    {
        return dummyTile.quantityDropped(meta, fortune);
    }

    @Override
    public void onNeighborBlockChange(World world, int x, int y, int z, int side)
    {
        inject(world, x, y, z);
        getTile(world, x, y, z).onNeighborChanged();
        eject();
    }

    @Override
    public boolean onBlockActivated(World world, int x, int y, int z, EntityPlayer player, int side, float hitX, float hitY, float hitZ)
    {
        inject(world, x, y, z);
        boolean value = getTile(world, x, y, z).activate(player, side, new Vector3(hitX, hitY, hitZ));
        eject();
        return value;
    }

    @Override
    public void updateTick(World world, int x, int y, int z, Random par5Random)
    {
        inject(world, x, y, z);
        getTile(world, x, y, z).updateEntity();
        eject();
    }

    @Override
    public void onEntityCollidedWithBlock(World world, int x, int y, int z, Entity entity)
    {
        inject(world, x, y, z);
        getTile(world, x, y, z).collide(entity);
        eject();
    }

    @Override
    public void addCollisionBoxesToList(World world, int x, int y, int z, AxisAlignedBB aabb, List list, Entity entity)
    {
        inject(world, x, y, z);

        Iterable<Cuboid> bounds = getTile(world, x, y, z).getCollisionBoxes(aabb != null ? new Cuboid(aabb).translate(new Vector3(x, y, z).invert()) : null, entity);

        if (bounds != null)
        {
            for (Cuboid cuboid : bounds)
            {
                list.add(cuboid.clone().translate(new Vector3(x, y, z)).toAABB());
            }
        }

        eject();
    }

    @SideOnly(Side.CLIENT)
    @Override
    public AxisAlignedBB getSelectedBoundingBoxFromPool(World world, int x, int y, int z)
    {
        inject(world, x, y, z);
        Cuboid value = getTile(world, x, y, z).getSelectBounds().clone().translate(new Vector3(x, y, z));
        eject();
        return value.toAABB();
    }

    @Override
    public AxisAlignedBB getCollisionBoundingBoxFromPool(World world, int x, int y, int z)
    {
        inject(world, x, y, z);
        Cuboid value = getTile(world, x, y, z).getCollisionBounds().clone().translate(new Vector3(x, y, z));
        eject();
        return value.toAABB();
    }

    @Override
    public boolean shouldSideBeRendered(IBlockAccess access, int x, int y, int z, int side)
    {
        return dummyTile.shouldSideBeRendered(access, x, y, z, side);
    }

    @Override
    public boolean isBlockSolid(IBlockAccess access, int x, int y, int z, int side)
    {
        inject(access, x, y, z);
        boolean value = getTile(access, x, y, z).isSolid(access, side);
        eject();
        return value;
    }

    @Override
    public int getLightValue(IBlockAccess access, int x, int y, int z)
    {
        int value = 0;

        if (access != null)
        {
            inject(access, x, y, z);
            value = getTile(access, x, y, z).getLightValue(access);
            eject();
        }
        return value;
    }

    @Override
    public boolean hasComparatorInputOverride()
    {
        return dummyTile instanceof IComparatorInputOverride;
    }

    @Override
    public boolean isOpaqueCube()
    {
        if (dummyTile == null)
        {
            return true;
        }

        return dummyTile.isOpaqueCube;
    }

    @Override
    public boolean renderAsNormalBlock()
    {
        return dummyTile.normalRender;
    }

    @SideOnly(Side.CLIENT)
    @Override
    public int getRenderType()
    {
        return BlockRenderingHandler.ID;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public Icon getBlockTexture(IBlockAccess access, int x, int y, int z, int side)
    {
        inject(access, x, y, z);
        Icon value = getTile(access, x, y, z).getIcon(access, side);
        eject();
        return value;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public Icon getIcon(int side, int meta)
    {
        return dummyTile.getIcon(side, meta);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerIcons(IconRegister iconRegister)
    {
        dummyTile.registerIcons(iconRegister);
    }

    @Override
    public ItemStack getPickBlock(MovingObjectPosition target, World world, int x, int y, int z)
    {
        inject(world, x, y, z);
        ItemStack value = getTile(world, x, y, z).getPickBlock(target);
        eject();
        return value;
    }

    @Override
    public ArrayList<ItemStack> getBlockDropped(World world, int x, int y, int z, int metadata, int fortune)
    {
        inject(world, x, y, z);
        ArrayList<ItemStack> value = getTile(world, x, y, z).getDrops(metadata, fortune);
        eject();
        return value != null ? value : new ArrayList<ItemStack>();
    }

    @Override
    protected void dropBlockAsItem_do(World world, int x, int y, int z, ItemStack itemStack)
    {
        if (!world.isRemote && world.getGameRules().getGameRuleBooleanValue("doTileDrops"))
        {
            InventoryUtility.dropItemStack(world, new Vector3(x, y, z), itemStack);
        }
    }

    @Override
    public void getSubBlocks(int id, CreativeTabs creativeTab, List list)
    {
        dummyTile.getSubBlocks(id, creativeTab, list);
    }

    /** Redstone interaction */
    @Override
    public boolean canProvidePower()
    {
        return dummyTile.canProvidePower;
    }

    @Override
    public int isProvidingWeakPower(IBlockAccess access, int x, int y, int z, int side)
    {
        inject(access, x, y, z);
        int value = getTile(access, x, y, z).getWeakRedstonePower(access, side);
        eject();
        return value;
    }

    @Override
    public int isProvidingStrongPower(IBlockAccess access, int x, int y, int z, int side)
    {
        inject(access, x, y, z);
        int value = getTile(access, x, y, z).getStrongRedstonePower(access, side);
        eject();
        return value;
    }

    @Override
    public int getRenderBlockPass()
    {
        return dummyTile.getRenderBlockPass();
    }

    @Override
    public int tickRate(World world)
    {
        return dummyTile.tickRate(world);
    }

}

File: src/main/scala/resonant/content/wrapper/BlockDummy.scala
package resonant.content.wrapper

import java.util.{ArrayList, List, Random}

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.block.{Block, ITileEntityProvider}
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.{Entity, EntityLivingBase}
import net.minecraft.item.{Item, ItemStack}
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.{AxisAlignedBB, IIcon, MovingObjectPosition}
import net.minecraft.world.{IBlockAccess, World}
import resonant.content.spatial.block.SpatialBlock
import resonant.lib.utility.inventory.InventoryUtility
import universalelectricity.core.transform.region.Cuboid
import universalelectricity.core.transform.vector.Vector3

class BlockDummy(val modPrefix: String, val defaultTab: CreativeTabs, val dummyTile: SpatialBlock) extends Block(dummyTile.material) with ITileEntityProvider
{
  dummyTile.domain = modPrefix
  setBlockName(modPrefix + dummyTile.name)
  setBlockTextureName(modPrefix + dummyTile.textureName)

  if (dummyTile.creativeTab != null)
  {
    setCreativeTab(dummyTile.creativeTab)
  }
  else
  {
    setCreativeTab(defaultTab)
  }

  dummyTile.bounds.setBounds(this)
  opaque = isOpaqueCube
  setLightOpacity(if (isOpaqueCube) 255 else 0)
  setHardness(dummyTile.blockHardness)
  setResistance(dummyTile.blockResistance)
  setTickRandomly(dummyTile.tickRandomly)
  setStepSound(dummyTile.stepSound)

  /**
   * Injects and ejects data from the TileEntity.
   */
  def inject(access: IBlockAccess, x: Int, y: Int, z: Int)
  {
    if (access.isInstanceOf[World])
    {
      dummyTile.world(access.asInstanceOf[World])
    }

    dummyTile._access = access
    dummyTile.xCoord = x
    dummyTile.yCoord = y
    dummyTile.zCoord = z

    val tile: TileEntity = access.getTileEntity(x, y, z)

    if (tile.isInstanceOf[SpatialBlock])
    {
      (tile.asInstanceOf[SpatialBlock]).block = this
    }
  }

  def eject()
  {
    dummyTile.world(null)
    dummyTile.xCoord = 0
    dummyTile.yCoord = 0
    dummyTile.zCoord = 0
  }

  def getTile(world: IBlockAccess, x: Int, y: Int, z: Int): SpatialBlock =
  {
    val tile: TileEntity = world.getTileEntity(x, y, z)
    if (tile.isInstanceOf[SpatialBlock])
    {
      return tile.asInstanceOf[SpatialBlock]
    }
    return dummyTile
  }

  override def hasTileEntity(metadata: Int): Boolean =
  {
    return dummyTile.tile != null
  }

  def createNewTileEntity(var1: World, var2: Int): TileEntity =
  {
    try
    {
      return dummyTile.tile.getClass.newInstance
    }
    catch
      {
        case e: Exception =>
        {
          e.printStackTrace
        }
      }
    return null
  }

  override def fillWithRain(world: World, x: Int, y: Int, z: Int)
  {
    inject(world, x, y, z)
    getTile(world, x, y, z).onFillRain
    eject
  }

  override def onBlockClicked(world: World, x: Int, y: Int, z: Int, player: EntityPlayer)
  {
    inject(world, x, y, z)
    getTile(world, x, y, z).click(player)
    eject
  }

  override def onBlockAdded(world: World, x: Int, y: Int, z: Int)
  {
    inject(world, x, y, z)
    getTile(world, x, y, z).onAdded
    eject
  }

  override def onBlockPlacedBy(world: World, x: Int, y: Int, z: Int, entityLiving: EntityLivingBase, itemStack: ItemStack)
  {
    inject(world, x, y, z)
    getTile(world, x, y, z).onPlaced(entityLiving, itemStack)
    eject
  }

  override def breakBlock(world: World, x: Int, y: Int, z: Int, block: Block, par6: Int)
  {
    inject(world, x, y, z)
    getTile(world, x, y, z).onRemove(block, par6)
    eject
    super.breakBlock(world, x, y, z, block, par6)
  }

  override def quantityDropped(meta: Int, fortune: Int, random: Random): Int =
  {
    return dummyTile.quantityDropped(meta, fortune)
  }

  override def onNeighborBlockChange(world: World, x: Int, y: Int, z: Int, block: Block)
  {
    inject(world, x, y, z)
    getTile(world, x, y, z).onNeighborChanged(block)
    eject
  }

  override def onBlockActivated(world: World, x: Int, y: Int, z: Int, player: EntityPlayer, side: Int, hitX: Float, hitY: Float, hitZ: Float): Boolean =
  {
    inject(world, x, y, z)
    val value: Boolean = getTile(world, x, y, z).activate(player, side, new Vector3(hitX, hitY, hitZ))
    eject
    return value
  }

  override def updateTick(world: World, x: Int, y: Int, z: Int, par5Random: Random)
  {
    inject(world, x, y, z)
    getTile(world, x, y, z).blockUpdate()
    eject
  }

  override def onEntityCollidedWithBlock(world: World, x: Int, y: Int, z: Int, entity: Entity)
  {
    inject(world, x, y, z)
    getTile(world, x, y, z).collide(entity)
    eject
  }

  override def addCollisionBoxesToList(world: World, x: Int, y: Int, z: Int, aabb: AxisAlignedBB, list: List[_], entity: Entity)
  {
    def add[T](list: java.util.List[T], value: Any) = list.add(value.asInstanceOf[T])

    inject(world, x, y, z)
    val bounds: Iterable[Cuboid] = getTile(world, x, y, z).getCollisionBoxes(if (aabb != null) new Cuboid(aabb) + (new Vector3(x, y, z) * -1) else null, entity)

    if (bounds != null)
    {
      for (cuboid <- bounds)
      {
        add(list, (cuboid + new Vector3(x, y, z)).toAABB)
      }
    }

    eject
  }

  @SideOnly(Side.CLIENT) override def getSelectedBoundingBoxFromPool(world: World, x: Int, y: Int, z: Int): AxisAlignedBB =
  {
    inject(world, x, y, z)
    val value: Cuboid = getTile(world, x, y, z).getSelectBounds.clone + new Vector3(x, y, z)
    eject
    return value.toAABB
  }

  override def getCollisionBoundingBoxFromPool(world: World, x: Int, y: Int, z: Int): AxisAlignedBB =
  {
    inject(world, x, y, z)
    val value: Cuboid = getTile(world, x, y, z).getCollisionBounds.clone + new Vector3(x, y, z)
    eject
    return value.toAABB
  }

  override def shouldSideBeRendered(access: IBlockAccess, x: Int, y: Int, z: Int, side: Int): Boolean =
  {
    return dummyTile.shouldSideBeRendered(access, x, y, z, side)
  }

  override def isBlockSolid(access: IBlockAccess, x: Int, y: Int, z: Int, side: Int): Boolean =
  {
    inject(access, x, y, z)
    val value: Boolean = getTile(access, x, y, z).isSolid(access, side)
    eject
    return value
  }

  override def getLightValue(access: IBlockAccess, x: Int, y: Int, z: Int): Int =
  {
    var value: Int = 0
    if (access != null)
    {
      inject(access, x, y, z)
      value = getTile(access, x, y, z).getLightValue(access)
      eject
    }
    return value
  }

  override def hasComparatorInputOverride: Boolean =
  {
    return dummyTile.isInstanceOf[SpatialBlock.IComparatorInputOverride]
  }

  override def isOpaqueCube: Boolean =
  {
    if (dummyTile == null)
    {
      return true
    }
    return dummyTile.isOpaqueCube
  }

  override def renderAsNormalBlock: Boolean =
  {
    return dummyTile.normalRender
  }

  @SideOnly(Side.CLIENT)
  override def getRenderType: Int =
  {
    return BlockRenderHandler.ID
  }

  @SideOnly(Side.CLIENT)
  override def getIcon(access: IBlockAccess, x: Int, y: Int, z: Int, side: Int): IIcon =
  {
    inject(access, x, y, z)
    val value: IIcon = getTile(access, x, y, z).getIcon(access, side)
    eject
    return value
  }

  @SideOnly(Side.CLIENT)
  override def getIcon(side: Int, meta: Int): IIcon =
  {
    return dummyTile.getIcon(side, meta)
  }

  @SideOnly(Side.CLIENT)
  override def registerBlockIcons(iconRegister: IIconRegister)
  {
    dummyTile.registerIcons(iconRegister)
  }

  @SideOnly(Side.CLIENT)
  override def colorMultiplier(access: IBlockAccess, x: Int, y: Int, z: Int): Int =
  {
    inject(access, x, y, z)
    val value = getTile(access, x, y, z).colorMultiplier()
    eject()
    return value
  }

  override def getPickBlock(target: MovingObjectPosition, world: World, x: Int, y: Int, z: Int): ItemStack =
  {
    inject(world, x, y, z)
    val value: ItemStack = getTile(world, x, y, z).getPickBlock(target)
    eject()
    return value
  }

  override def getDrops(world: World, x: Int, y: Int, z: Int, metadata: Int, fortune: Int): ArrayList[ItemStack] =
  {
    inject(world, x, y, z)
    val value: ArrayList[ItemStack] = getTile(world, x, y, z).getDrops(metadata, fortune)
    eject()
    return if (value != null) value else new ArrayList[ItemStack]
  }

  protected override def dropBlockAsItem(world: World, x: Int, y: Int, z: Int, itemStack: ItemStack)
  {
    if (!world.isRemote && world.getGameRules.getGameRuleBooleanValue("doTileDrops"))
    {
      InventoryUtility.dropItemStack(world, new Vector3(x, y, z), itemStack)
    }
  }

  override def getSubBlocks(item: Item, creativeTabs: CreativeTabs, list: List[_])
  {
    dummyTile.getSubBlocks(item, creativeTabs, list)
  }

  /**
   * Redstone interaction
   */
  override def canProvidePower: Boolean =
  {
    return dummyTile.canProvidePower
  }

  override def isProvidingWeakPower(access: IBlockAccess, x: Int, y: Int, z: Int, side: Int): Int =
  {
    inject(access, x, y, z)
    val value: Int = getTile(access, x, y, z).getWeakRedstonePower(access, side)
    eject
    return value
  }

  override def isProvidingStrongPower(access: IBlockAccess, x: Int, y: Int, z: Int, side: Int): Int =
  {
    inject(access, x, y, z)
    val value: Int = getTile(access, x, y, z).getStrongRedstonePower(access, side)
    eject
    return value
  }

  override def getRenderBlockPass: Int =
  {
    return dummyTile.getRenderBlockPass
  }

  override def tickRate(world: World): Int =
  {
    return dummyTile.tickRate(world)
  }
}